Documentation/ExtensionArchitecture/Extbase/Reference/FileUpload.rst

```diff
@@ -0,0 +1,489 @@
+`Torben Hansen's EXT:extbase-upload <https://github.com/derhansen/extbase_upload>`__
+repository.
+
+The general idea is to use PHP attributes within the Extbase Model, and for the upload
+use a custom ViewHelper:
+
+..  literalinclude:: _FileUpload/_BlogEnhanced.php
+    :caption: EXT:my_extension/Classes/Domain/Model/Blog.php, using FileUpload attributes
+
+..  literalinclude:: _FileUpload/_Upload.html
+    :caption: EXT:my_extension/Resources/Private/Templates/Blog/New.html
+
+You can actually easily also allow to remove already uploaded files:
+
+..  literalinclude:: _FileUpload/_UploadDelete.html
+    :caption: EXT:my_extension/Resources/Private/Templates/Blog/New.html
+
+The controller action part with persisting the data needs no further custom code,
+Extbase can automatically do all the Domain Model handling on its own. The TCA can
+also stay the same as configured for simply read-access to a Domain Model.
+
+The actual file upload processing is performed just before Extbase persists a
+Domain Model Entity, so only if the full validation has taken place and no errors
+occurred. This means, if any error occurs, a user will have to re-upload a file.
+
+The implementation is done like this to prevent stale temporary files that would
+need cleanup or could raise issues with Denial of Service.
+
+
+..  _extbase_fileupload_attribute:
+
+Reference for the :php:`FileUpload` PHP attribute
+-------------------------------------------------
+
+File uploads can be validated by the following rules:
+
+*   minimum and maximum file count
+*   minimum and maximum file size
+*   allowed MIME types
+*   image dimensions (for image uploads)
+
+Additionally, it is ensured, that the filename given by the client is valid,
+meaning that no invalid characters (null-bytes) are added and that the file
+does not contain an invalid file extension. The API has support for custom
+validators, which can be created on demand.
+
+To avoid complexity and maintain data integrity, a file upload is only
+processed if the validation of all properties of a domain model is successful.
+In this first implementation, file uploads are not persisted/cached temporarily,
+so this means in any case of a validation failure ("normal" validators and file upload
+validation) a file upload must be performed again by users.
+
+Possible future enhancements of this functionality could enhance the existing
+`#[FileUpload]` attribute/annotation with configuration like a temporary storage
+location, or specifying additional custom validators (which can be done via the PHP-API as
+described below)
+
+..  _extbase_fileupload_attribute_configuration:
+File upload configuration with the `FileUpload` attribute
+---------------------------------------------------------
+
+File upload for a property of a domain model can be configured using the
+newly introduced :php:`\TYPO3\CMS\Extbase\Annotation\FileUpload` attribute.
+
+Example:
+
+..  code-block:: php
+
+    #[FileUpload([
```

Comment:
Please surround it in a class (I assume it is a model) and add a caption for clarity.

Reply:
Good idea. Just taken from the changelog.

Reply:
done